src/cargo/core/compiler/fingerprint.rs

```diff
@@ -1458,9 +1455,15 @@ where
 
     for path in paths {
         let path = path.as_ref();
-        let path_mtime = match paths::mtime(path) {
-            Ok(mtime) => mtime,
-            Err(..) => return Some(StaleFile::Missing(path.to_path_buf())),
+        let path_mtime = match mtime_cache.entry(path.to_path_buf()) {
```

Comment:
This is the most worrisome part here to me, using the mtime cache for all calls to `paths::mtime`. My main worry is how this interacts with edits during the build itself. Can you confirm that `find_stale_file` here is basically only called during "startup" of Cargo? I think it's fine to, basically during the preparation phase of a build, assume that everything happens instantaneously. 

I *think* that's the case here already since it's the only phase which should have `&mut Context`, but would be worth double-checking.

Reply:
Also you were mentioning that this deduplicates sysroot dependencies, but that's the only mtime calls that are being reduced, right?

Reply:
Yes, it is all done at the beginning.

It can also deduplicate other paths as well.  For example, when doing `cargo build --tests` then `main.rs` will show up twice (once as a dep of the bin executable, and once as the bin unittest).  There are a variety of other situations (using the same modules from different targets, `include` shenanigans, etc.).  I'm fairly confident they should all be fine.

Reply:
Ok sounds good to me!

One step we could also do is to store `Option<HashMap>` and unwrap it here, and then after the build preparation is finished we set it to `None` to basically internally assert we don't access it again. I don't think that's necessary at this time, though.